Enabled filling color by value for :class:`.OpenGLSurface`, replaced ``colors`` keyword argument of :meth:`.Surface.set_fill_by_value`  with ``colorscale``

alembcke opened the pull request:
<!-- Thank you for contributing to Manim! Learn more about the process in our contributing guidelines: https://docs.manim.community/en/latest/contributing.html -->

This pull requests closes #2173 .

## Overview: What does this pull request change?
<!-- If there is more information than the PR title that should be added to our release changelog, add it in the following changelog section. This is optional, but recommended for larger pull requests. -->
<!--changelog-start-->
Enables coloring `OpenGLSurface` gradient colors by z-value, by passing in a list of colors and an axes when a `OpenGLSurface` is initialized.  The method `OpenGLSurface.set_fill_by_value()` has been deprecated, as it never worked properly for `OpenGLSurface`.
<!--changelog-end-->

## Motivation and Explanation: Why and how do your changes improve the library?
<!-- Optional for bugfixes, small enhancements, and documentation-related PRs. Otherwise, please give a short reasoning for your changes. -->
Adds similar functionality as `Surface.set_fill_by_value()`, but for `OpenGLSurface`.  Here is an example:

![Test_ManimCE_v0 11 0](https://user-images.githubusercontent.com/79891353/136728799-32b2f54c-69d7-4e4e-9192-9f07243b8378.png)

And the code used to generate this example:

```py
from manim import *
from manim.opengl import *

import numpy as np


class Test(ThreeDScene):
    def construct(self):
        resolution_fa = 100
        self.set_camera_orientation(phi=75 * DEGREES, theta=-60 * DEGREES)

        axes = ThreeDAxes(
                x_range=(-3, 3, 1),
                y_range=(-3, 3, 1),
                z_range=(-5, 5, 1)
                )

        def param_trig(u, v):
            x = u
            y = v
            z = 2 * np.sin(x) + 2 * np.cos(y)
            return z

        trig_plane = OpenGLSurface(
            lambda x, y: axes.c2p(x, y, param_trig(x, y)),
            resolution=(resolution_fa, resolution_fa),
            u_range = (-3, 3),
            v_range = (-3, 3),
            axes = axes,
            color = [BLUE, GREEN, YELLOW, ORANGE, RED],
            opacity = 0.8
        )

        self.add(axes, trig_plane)

        self.interactive_embed()
```

## Links to added or changed documentation pages
<!-- Please add links to the affected documentation pages (edit the description after opening the PR). The link to the documentation for your PR is https://manimce--####.org.readthedocs.build/en/####/, where #### represents the PR number. -->


## Further Information and Comments
<!-- If applicable, put further comments for the reviewers here. -->
The method `OpenGLSurface.set_fill_by_value()` has been deprecated as it never worked properly.  The same method worked for `Surface` because `Surface` is made up of many small squares lined up next to each other, so it could color each square a different color and it would give a gradient effect.  But `OpenGLSurface` is made up of triangles, which are fused together to create a single `mobject`.  So to color a `OpenGLSurface` it is necessary to pass in the colors to the shaders, along with the vertex data at the time the surface is initialized.


<!-- Thank you again for contributing! Do not modify the lines below, they are for reviewers. -->
## Reviewer Checklist
- [x] The PR title is descriptive enough for the changelog, and the PR is labeled correctly
- [ ] If applicable: newly added non-private functions and classes have a docstring including a short summary and a PARAMETERS section
- [ ] If applicable: newly added functions and classes are tested

behackl commented on manim/mobject/types/opengl_surface.py:
```diff
@@ -6,12 +6,36 @@
 
 class OpenGLSurface(OpenGLMobject):
+    """Creates a Surface.
+    Parameters
+    ----------
+    uv_func :
+        The function that defines the surface.
+    u_range :
+        The range of the ``u`` variable: ``(u_min, u_max)``.
+    v_range :
+        The range of the ``v`` variable: ``(v_min, v_max)``.
+    resolution :
+        The number of samples taken of the surface.
+    axes :
+        Axes on which the surface is to be drawn. Optional
+        parameter used when coloring a surface by z-value.
+    color :
+        Color of the surface. Passing a list of colors and a
+        axes will color the surface by z-value. Passing a
+        list of tuples in the form `[(color, pivot)]` allows
+        user-defined pivots where the color transitions.
+    opacity :
+        Opacity of the surface.
```
The colons should be removed, this change has not been incorporated.

alembcke commented on manim/mobject/opengl_mobject.py:
```diff
@@ -81,7 +81,15 @@ def __init__(
         self.uniforms = getattr(self, "uniforms", {})
 
-        self.color = Color(color) if color else None
+        if isinstance(color, list):
+            self.color = []
+            for item in color:
+                if isinstance(item, tuple):
+                    self.color.append(Color(item[0]))
+                else:
+                    self.color.append(Color(item))
+        else:
+            self.color = Color(color) if color else None
```
Just to add to the above, the `else` clause keeps everything the way it was originally - which is that you would get an error if you passed in anything other than a color.  What was added here was just logic to parse out the color if you are passing in something compatible with setting the color by z-value.  So the code expands what can be passed in to accommodate setting the color by z-value.

behackl replied:
What if we make things a bit more explicit: check whether the passed value is a string or already a color first, otherwise check whether it is an iterable (not just a list) whose elements we can then assume to be either colors or tuples, and if thats also not true raise a ValueError?

alembcke replied:
I updated the code to this:

```py
if isinstance(color, (str, Color)):
    self.color = Color(color)
elif isinstance(color, Iterable):   
    self.color = []
    for item in color:
        if isinstance(item, tuple):
            self.color.append(Color(item[0]))
        else:
            self.color.append(Color(item))
else:
    raise ValueError("Invalid color type: " + str(color))
```

The nested `isinstance(item, tuple)` is still looking for a tuple because a string is also an iterable.  Or should I change the nested if/else to check first for a string or color and then iterable and then error if none of those?

alembcke replied:
I was able to get rid of this entirely by using a new argument, `colorscale`.

alembcke commented:
@behackl and @hydrobeam, I have been thinking a lot about the implementation of this enhancement and the concerns raised.  I think it would be best if I spent some time completely re-doing the implementation.  So with that in mind, I have some questions for you, or anyone else who is interested:

1. I agree that this should be moved to `coordinate_systems.py` within `ThreeDAxes` as the method `~.plot_surface()`.  I will switch between `Surface` and `OpenGLSurface` based on which renderer is being used.  Is everyone happy with that idea?

2. How should colors and pivots be set by the user?  I have gone with a list of tuples approach, but would it be best to separate these out into two variables?  There are other alternatives, for example Plotly has the user pass in a colorscale using the argument `colorscale` and then the pivots are set using `cmin`, `cmax` and `cmid`.  Personally, I prefer to allow the user to set each pivot for maximum control, but am open to other ideas.

3. The last part of the above question raises another one, should there be a new variable for colorscales of not just surfaces, but any `MObject` or `VMObject`?  If so, should it be called `colorscale` or something else?

4. Should I also look to change the implementation of `Surface.set_fill_by_value()` to align it with what we are doing here?  Making it a part of `~.plot_surface()` and deprecate `~.set_fill_by_value()`?

For me, the key functionality is the ability to apply a colorscale by z-value to surfaces and to be able to have below zero be one color and above zero be another color (I am making finance videos, so want to show below zero as red for loss and above zero as green for profit).

Feel free to add anything I may have missed.

alembcke commented:
I have made the below changes:

1.  Added `axes.plot_surface()` which switches between Cairo and OpenGL versions of `Surface` and includes `colorscale` as an argument.

2.  Modified `OpenGLSurface.set_color_by_value()` and `Surface.set_color_by_value()` to use `colorscale` instead of `color` as the input.  I kept a list of lists/tuples as the method for manually setting pivot points, as I think that makes the most sense.

3.  Added tests for the Cairo version of `axes.plot_surface()`.  I checked for the OpenGL file to add the tests, but it doesn't exist (nor do any of the graphical tests) and the OpenGL tests look different from the Cairo test, so wasn't sure how to implement this.

I hope this resolves most of the issues raised.  One issue that I could not solve was not making `axes` an attribute of `OpenGLSurface`.  The reason being that the way the shader data is populated is that `OpenGLSurface.get_shader_data()` is called from the parent `Mobject` class.  But let me know if I missed something.

alembcke commented on manim/mobject/types/opengl_surface.py:
```diff
@@ -6,12 +6,37 @@
+        The range of the ``v`` variable: ``(v_min, v_max)``.
+    resolution
+        The number of samples taken of the surface.
+    axes
+        Axes on which the surface is to be drawn. Optional
+        parameter used when coloring a surface by z-value.
```
I have implemented `axes.plot_surface()`, but am unable to find a way to not make `axes` an attribute of `OpenGLSurface`.

alembcke commented on manim/mobject/opengl_mobject.py:
```diff
@@ -1327,7 +1335,15 @@ def set_color(self, color, opacity=None, recurse=True):
+                    else:
+                        self.color.append(Color(item))
+            else:
+                self.color = Color(color)
```
Also resolved via the new variable `colorscale`.

alembcke commented on manim/mobject/types/opengl_surface.py:
```diff
@@ -206,16 +233,68 @@ def get_shader_data(self):
             shader_data["du_point"] = du_points
             shader_data["dv_point"] = dv_points
-        self.fill_in_shader_color_info(shader_data)
+            if isinstance(self.color, list):
```
Also got rid of this by using the new variable `colorscale`.

alembcke commented:
Rebased to account for PR #2476 and cleaned up the docstring.

behackl reviewed (approved):
I've added one commit properly deprecating the `color` parameter `Surface.set_fill_by_value` to avoid introducing a hard breaking change, please cross-check! Otherwise this still looks fine to me, if someone using the OpenGL renderer a bit more than myself could try to run this locally too it would be much appreciated.

Thanks for your contribution!
